addons/mrp_account/tests/test_bom_price.py

```diff
@@ -254,3 +254,85 @@ def test_02_compute_byproduct_price(self):
+        self.product_category = self.env['product.category'].with_company(self.company_b).create({
+            'name': 'Test Category',
+            'property_cost_method': 'average',
+        })
+
+        # Product for Company B
+        self.main_product = self.env['product.product'].create({
+            'name': 'Main Product',
+            'type': 'product',
+            'standard_price': 101.11,
+            'company_id': self.company_b.id,
+            'categ_id': self.product_category.id,
+        })
+
+        # Create additional product for the BOM
+        self.product_b = self.env['product.product'].with_company(self.company_b).create({
+            'name': 'Product B',
+            'type': 'product',
+            'standard_price': 100,
+            'company_id': self.company_b.id,
+            'categ_id': self.product_category.id,
+        })
+
+        # Create Bill of Materials (BoM) for the main product
+        self.bom = self.env['mrp.bom'].with_company(self.company_b).create({
+            'product_id': self.main_product.id,
+            'product_tmpl_id': self.main_product.product_tmpl_id.id,
+            'product_qty': 1.0,
+            'type': 'phantom',
+            'bom_line_ids': [(0, 0, {'product_id': self.product_b.id, 'product_qty': 1})],
+        })
+
+        # Create stock moves for Product B
+        self.stock_moves = self.env['stock.move'].create([
+            {
+                'name': 'Move 1 - Product B',
+                'product_id': self.product_b.id,
+                'company_id': self.company_b.id,
+                'product_uom_qty': 5,
+                'product_uom': self.product_b.uom_id.id,
+                'bom_line_id': self.bom.bom_line_ids[0].id,
+                'location_id': self.env.ref('stock.stock_location_stock').id,
+                'location_dest_id': self.env.ref('stock.stock_location_customers').id,
+                'state': 'done',
+            },
+        ])
+
+        # Create Sale Order
+        self.sale_order = self.env['sale.order'].create({
+            'partner_id': self.env.ref('base.res_partner_1').id,
+            'company_id': self.company_b.id,
+        })
+
+        # Create Sale Order Line with main_product and link stock_moves
+        self.sale_order_line = self.env['sale.order.line'].create({
+            'order_id': self.sale_order.id,
+            'product_id': self.main_product.id,
+            'product_uom_qty': 1,
+            'product_uom': self.main_product.uom_id.id,
+            'company_id': self.company_b.id,
+            'move_ids': self.stock_moves.ids,
+        })
+
+        # Switch to Company A's context for the test
+        self.env.company = self.company_a
+
+        self.sale_order_line._compute_purchase_price()
+        purchase_price = self.sale_order_line.purchase_price
+
+        # Assert that the computed price comes from Company B
+        self.assertNotEqual(purchase_price, 0, "Price should not be 0 as the product exists in Company B")
```

Comment:
This assert is not necessary, you can instead make an assert to check that the product price in company A is 0